jax-stuff.md

```diff
@@ -150,36 +152,86 @@ mesh = jax.make_mesh((4, 2), ('X', 'Y'))
 
 def matmul(x, Win, Wout):
   hidden = jnp.einsum('bd,df->bf', x, Win)
-  hidden = jax.lax.with_sharding_constraint(hidden, jax.sharding.NamedSharding(mesh, P('x', 'y')))
+  hidden = jax.lax.with_sharding_constraint(hidden, jax.P('x', 'y'))
   return jnp.einsum('bf,df->bd', hidden, Wout)
 ```
 
-This makes up like 60% of JAX parallel programming in the jit world, since it's our only way of intervening with the compiler. It's worth playing around with `with_sharding_constraint` in a Colab and getting a sense for how it works. When we write LLMs using `jax.jit`, 90% of what we do to control shardings is changing the input and output shardings (via `in_shardings` and `out_shardings`) and annotating intermediate tensors with `with_sharding_constraint` to ensure the correct comms are happening. For more jax.jit examples, [this is a great doc to read](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html).
+Looking at a profile and figuring out how to shard your intermediates because Shardy picked a non-ideal sharding is not a good user experience. Enter explicit mode!
+
+<h3 id="explicit-sharding-mode">Explicit sharding mode</h3>
+
+Auto mode leaves sharding decisions to the compiler. But what if you have a strong opinion about how you want your program sharded? With enough calls to jax.lax.with_sharding_constraint you can probably guide the compiler’s hand to make it do what you want. But “compiler tickling” is famously not a fun programming model. Where should you put the sharding constraints? You could put them on every single intermediate but that’s a lot of work and it’s also easy to make mistakes that way because there’s no way to check that the shardings make sense together. More commonly, people add just enough sharding annotations to constrain the compiler. But this is a slow iterative process. It’s hard to know ahead of time what XLA’s Shardy pass will do (it’s a whole-program optimization) so all you can do is add annotations, inspect XLA’s sharding choices to see what happened by looking at the profile, and repeat.
+
+To fix this we’ve come up with a different style of sharding programming we call “explicit sharding” or “sharding in types”. The idea is that sharding propagation happens at the JAX level at trace time. Each JAX operation has a sharding rule that takes the shardings of the op’s arguments and produces a sharding for the op’s result. For most operations these rules are simple and obvious because there’s only one reasonable choice. But for some operations it’s ambiguous how to shard the result in which case JAX throws a trace-time error and we ask the programmer to provide an `out_sharding` argument explicitly.
```

Comment:
nit: provide an example like "For most operations these rules are simple and obvious because there's only one reasonable choice (e.g. elementwise ops retain the same sharding)."

Reply:
Done